### PR TITLE
Have dispatched functions appear as normal Python functions when type-checking

### PR DESCRIPTION
**Pull request by eriknw**
This is an effort to make type-checkers happy. Without this, dispatch functions are of type `_dispatch`, which may complicate typing.

For now, the attributes, methods, and type of `_dispatch` objects are implementation details and may be treated as regular Python functions, which this PR makes explicit.

### CODE REVIEW

File: networkx/utils/backends.py [hunk 1]
```diff
@@ -1002,17 +1003,45 @@ def _restore_dispatch(name):
...
     def _dispatch(func=None, **kwargs):  # type: ignore[no-redef]
         if func is None:
             return partial(_dispatch, **kwargs)
         dispatched_func = _orig_dispatch(func, **kwargs)
         func.__doc__ = dispatched_func.__doc__
         return func
+
+elif TYPE_CHECKING:
+    # For type-checking, have dispatched functions show up as normal Python functions.
+    # These do have `backend=None` and `**backend_kwargs` added to the signature!
```
Comment (MridulS) [inline networkx/utils/backends.py:1023]: This means all the functions would need to add this to their type annotations right?

Comment (eriknw) [reply networkx/utils/backends.py:1023]: Yes, as is currently done in this PR. Should we not have the new arguments?

**Review by rlratzel — APPROVED** [review]
Looks fine to me.  This may be out-of-scope for this PR, but fixing the interactive `help` would be nice too:
```python
>>> networkx.__version__
'2.8.8'
>>> help(networkx.pagerank)
Help on function pagerank in module networkx.algorithms.link_analysis.pagerank_alg:

pagerank(G, alpha=0.85, personalization=None, max_iter=100, tol=1e-06, nstart=None, weight='weight', dangling=None)
    Returns the PageRank of the nodes in the graph.
...
```
vs.
```python
>>> networkx.__version__
'3.2.1'
>>> help(networkx.pagerank)
Help on _dispatch in module networkx.algorithms.link_analysis.pagerank_alg:

pagerank = <networkx.utils.backends._dispatch object>
    Returns the PageRank of the nodes in the graph.
...
```

**Comment by eriknw** [comment]
> fixing the interactive help would be nice too

This requires us to define `__repr__`. What should the repr and str of dispatchable objects be?

**Comment by dschult** [comment]
Could `__repr__` and `__str__` be the same as it was without the dispatchable decorator? 
- is there any reason to include more?
- are there side-effects from getting those from the wrapped function and attaching them to the wrapper?

**Comment by eriknw** [comment]
> Could `__repr__` and `__str__` be the same as it was without the dispatchable decorator?
> 
> * is there any reason to include more?
> * are there side-effects from getting those from the wrapped function and attaching them to the wrapper?

Here's what we have today.
Python REPR:
```python
>>> import networkx as nx
>>> print(str(nx.pagerank.orig_func))
<function pagerank at 0x7f038ebe4310>
>>> print(repr(nx.pagerank.orig_func))
<function pagerank at 0x7f038ebe4310>
>>> print(str(nx.pagerank))
<networkx.utils.backends._dispatchable object at 0x7f038ebe0eb0>
>>> print(repr(nx.pagerank))
<networkx.utils.backends._dispatchable object at 0x7f038ebe0eb0>
>>> help(nx.pagerank.orig_func)
Help on function pagerank in module networkx.algorithms.link_analysis.pagerank_alg:

pagerank(G, alpha=0.85, personalization=None, max_iter=100, tol=1e-06, nstart=None, weight='weight', dangling=None)
    Returns the PageRank of the nodes in the graph.
    ...

>>> help(nx.pagerank)
Help on _dispatchable in module networkx.algorithms.link_analysis.pagerank_alg:

pagerank = <networkx.utils.backends._dispatchable object>
    Returns the PageRank of the nodes in the graph.
    ...
```

IPython REPR:
```python
In [1]: import networkx as nx

In [2]: nx.pagerank.orig_func
Out[2]: <function networkx.algorithms.link_analysis.pagerank_alg.pagerank(G, alpha=0.85, personalization=None, max_iter=100, tol=1e-06, nstart=None, weight='weight', dangling=None)>

In [3]: nx.pagerank
Out[3]: <networkx.utils.backends._dispatchable at 0x7f384345aa70>

In [4]: nx.pagerank.orig_func?
Signature:
nx.pagerank.orig_func(
    G,
    alpha=0.85,
    personalization=None,
    max_iter=100,
    tol=1e-06,
    nstart=None,
    weight='weight',
    dangling=None,
)
Docstring:
Returns the PageRank of the nodes in the graph.
...

In [5]: nx.pagerank?
Signature:
nx.pagerank(
    G,
    alpha=0.85,
    personalization=None,
    max_iter=100,
    tol=1e-06,
    nstart=None,
    weight='weight',
    dangling=None,
    *,
    backend=None,
    **backend_kwargs,
)
Call signature: nx.pagerank(*args, backend=None, **kwargs)
Type:           _dispatchable
String form:    <networkx.utils.backends._dispatchable object at 0x7f384345aa70>
File:           ~/git/networkx/networkx/algorithms/link_analysis/pagerank_alg.py
Docstring:
Returns the PageRank of the nodes in the graph.
```

In other words, it's tricky to get consistent/desirable/familiar behavior just by using dunder methods on `_dispatchable` when the networkx function is an instance of `_dispatchable`.

We can use `argmap` to turn a `_dispatchable` back into a normal function:
```python
In [2]: def do_nothing():
   ...:     pass
   ...:

In [3]: pagerank = nx.utils.argmap(do_nothing)(nx.pagerank)

In [4]: pagerank
Out[4]: <function networkx.algorithms.link_analysis.pagerank_alg.pagerank(G, alpha=0.85, personalization=None, max_iter=100, tol=1e-06, nstart=None, weight='weight', dangling=None, *, backend=None, **backend_kwargs)>
```
Note how this correctly has `backend=None, **backend_kwargs`. I'll note that I don't understand how argmap works.

Should we always wrap `_dispatchable` with `argmap` (inside `_dispatchable.__new__`) to make it a normal function? This should also make type-checking work without the workarounds in this PR.

**Comment by eriknw** [comment]
Closing. This was superceded by https://github.com/networkx/networkx/pull/7298, which was merged.

**Pull request closed without merging**